x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/AuthenticationService.java

```diff
@@ -462,11 +462,15 @@ private void consumeToken(AuthenticationToken token) {
                                 // the user was not authenticated, call this so we can audit the correct event
                                 request.realmAuthenticationFailed(authenticationToken, realm.name());
                                 if (result.getStatus() == AuthenticationResult.Status.TERMINATE) {
-                                    logger.info("Authentication of [{}] was terminated by realm [{}] - {}",
-                                        authenticationToken.principal(), realm.name(), result.getMessage());
-                                    Exception e = (result.getException() != null) ? result.getException()
-                                        : Exceptions.authenticationError(result.getMessage());
-                                    userListener.onFailure(e);
+                                    if (result.getException() != null) {
+                                        logger.info(new ParameterizedMessage("Authentication of [{}] was exceptionally terminated by " +
+                                                        "realm [{}] - {}", authenticationToken.principal(), realm.name(),
+                                                        result.getMessage()), result.getException());
+                                    } else {
+                                        logger.info("Authentication of [{}] was gracefully terminated by realm [{}] - {}",
+                                                authenticationToken.principal(), realm.name(), result.getMessage());
+                                    }
+                                    userListener.onFailure(result.getException());
```

Comment:
It seems weird to have an `onFailure` with a possibly null exception. I get that it's handled correctly below, but can't it be handled here and be more idiomatic?

Reply:
Subjectively, I don't see a better way.
The only way we can signal that this is a case of `authenticationFailed` as opposed to a case of `exceptionProcessingRequest` is through the value of the `onFailure` handler. Instead of the null thing, we can fill in a static exception value that codes the `null` value, i.e.
```
--- a/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/AuthenticationService.java
+++ b/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/AuthenticationService.java
@@ -77,6 +77,7 @@ public class AuthenticationService {
         Setting.intSetting("xpack.security.authc.success_cache.size", 10000, Property.NodeScope);
     private static final Setting<TimeValue> SUCCESS_AUTH_CACHE_EXPIRE_AFTER_ACCESS =
         Setting.timeSetting("xpack.security.authc.success_cache.expire_after_access", TimeValue.timeValueHours(1L), Property.NodeScope);
+    private static final Exception AUTHN_TERMINATED_NO_EXCEPTION = Exceptions.authenticationError("Authn terminated");
     private static final Logger logger = LogManager.getLogger(AuthenticationService.class);

     private final Realms realms;
@@ -469,11 +470,12 @@ public class AuthenticationService {
                                         logger.info(new ParameterizedMessage(
                                                 "Authentication of [{}] was terminated by realm [{}] - {}",
                                                 authenticationToken.principal(), realm.name(), result.getMessage()), result.getException());
+                                        userListener.onFailure(result.getException());
                                     } else {
                                         logger.info("Authentication of [{}] was terminated by realm [{}] - {}",
                                                 authenticationToken.principal(), realm.name(), result.getMessage());
+                                        userListener.onFailure(AUTHN_TERMINATED_NO_EXCEPTION);
                                     }
-                                    userListener.onFailure(result.getException());
                                 } else {
                                     if (result.getMessage() != null) {
                                         messages.put(realm, new Tuple<>(result.getMessage(), result.getException()));
@@ -496,7 +498,7 @@ public class AuthenticationService {
                     new IteratingActionListener<>(ContextPreservingActionListener.wrapPreservingContext(ActionListener.wrap(
                         (user) -> consumeUser(user, messages),
                         (e) -> {
-                            if (e != null) {
+                            if (e != AUTHN_TERMINATED_NO_EXCEPTION) {
                                 listener.onFailure(request.exceptionProcessingRequest(e, token));
                             } else {
                                 listener.onFailure(request.authenticationFailed(token));
```

I personally prefer the `null` value to flag values like these ones.

If you or @ywangd  think otherwise, I'll go with that.

Reply:
I would choose `null` over the static exception as well. Afterall, the original error from ReservedRealm is `null`.

Alternatively, if we wanna have some brainstorming here, we could give the `IteratingActionListener` a non-default `iterationPredicate` so it can terminate earlier even with a `null` user response, e.g. something like the follows:

```java
final AtomicBoolean terminated = new AtomicBoolean(false);
final BiConsumer<Realm, ActionListener<User>> realmAuthenticatingConsumer = (realm, userListener) -> {
   ...
   if (request.getException() == null) {
        logger.info(...);
        terminated.set(true);
        userListener.onResponse(null)
   } else {
        logger.info(...);
        userListener.onFailure(request.getException());
   }
   ...
};
...
final IteratingActionListener<User, Realm> authenticatingListener =
    new IteratingActionListener<>(..., 
        Function.identity(), 
        user -> Objects.isNull(user) && terminated.get() == false);
```

Yet another option is to have a static `NOT_AUTHENTICATED_USER` object as a static field in `User` class and use it to signal earlier termination:
```java
   if (request.getException() == null) {
        logger.info(...);
        userListener.onResponse(User.NOT_AUTHENTICATED_USER)
   } else {
        logger.info(...);
        userListener.onFailure(request.getException());
   }
```
In `consumeUser()`, we will then need to check both `null` and this object, i.e. 
```java
if (user == null || user == User.NOT_AUTHENTICATED_USER) { ... }
```
Or we could give `IteratingActionListener` a transform function of `response -> null ? User.NOT_AUTHENTICATED_USER : response` so that `consumeUser()` can drop the null check.

Both of the alternatives reuse the logic in `consumeUser`, but add complexity elsewhere. I don't think they are better, just different choices. I personally would probably go with the static `User` field approach, but the `null` exception approach is fine with me.